source/extensions/access_loggers/open_telemetry/grpc_access_log_impl.cc

```diff
@@ -44,6 +35,30 @@ opentelemetry::proto::common::v1::KeyValue getStringKeyValue(const std::string&
 
 } // namespace
 
+GrpcAccessLoggerImpl::GrpcAccessLoggerImpl(
+    const Grpc::RawAsyncClientSharedPtr& client,
+    const envoy::extensions::access_loggers::open_telemetry::v3::OpenTelemetryAccessLogConfig&
+        config,
+    Event::Dispatcher& dispatcher, const LocalInfo::LocalInfo& local_info, Stats::Scope& scope)
+    : GrpcAccessLogger(
+          config.common_config(), dispatcher, scope, std::nullopt,
+          std::make_unique<Common::UnaryGrpcAccessLogClient<ExportLogsServiceRequest,
+                                                            ExportLogsServiceResponse>>(
+              client,
+              *Protobuf::DescriptorPool::generated_pool()->FindMethodByName(
+                  "opentelemetry.proto.collector.logs.v1.LogsService.Export"),
+              GrpcCommon::optionalRetryPolicy(config.common_config()),
+              [this, &dispatcher]() {
+                // It will be deleted in the callbacks.
+                auto* callback = new OTelLogRequestCallbacks(dispatcher, this->stats_,
+                                                             this->batched_log_entries_);
+                this->batched_log_entries_ = 0;
+                return callback;
```

Comment:
Rather then create a self-management request callbacks object, I prefer to manage all these pending requests in the unary client directly. And the unary client should provide a way to let the specific implementation to provide a callbacks.
Then things would be simpler and the lifetime of these objects would be more clear.

Reply:
Hi @wbpcode , could you please explain which class owns what and how the deletion works in your mind? Thanks

Reply:
I will assume the unary client will maintains a set of pending requests in a `flat_hash_map`. The request pointer (returned by the `send()`) will be used as the key. And the value is a simple wrapper that like current `OTelLogRequestCallbacks`. The wrapper will delegate the callback calling to the unary client to remove itself from the map and execute some logic that specified by the specific logger.
**Then when the unary client self is destroyed, it will cancel all pending requests to ensure no any further logic will be executed.** (This is the key difference)

Reply:
Thanks for your quick response and hard work. But I think the most important point still doesn't be resolved.

> Then when the unary client self is destroyed, it will cancel all pending requests to ensure no any further logic will be executed. (This is the key difference)

See, the pending requests will still be pending when the client is destroyed and the pending requests may use dangling reference to the callbacks objects or dangling reference to the client stats.

Reply:
Thanks!

You are talking about the potential issues: the ongoing stream access the deleted callback/stats owned under `GrpcAccessLoggerImpl`.

I think in theory it shouldn't happen because `GrpcAccessLoggerImpl` as `common::GrpcAccessLogger`owns the `AsyncClientImpl` , holding the `AsyncStream`. So when  `GrpcAccessLoggerImpl` is destructed, all the active streams should be destructed as well and there won't be callbacks in this case.

Reply:
I did a quick check to the source code of grpc client. Yeah, all active streams will be destroyed automatically. But seems you still need to call the cancel() explicitly to reset underlying http request, needn't it?
(Will do a deeper check tomorrow in case I miss something)

Reply:
When AsyncStreamImpl get deleted, it will reset the underlying http stream.

https://github.com/envoyproxy/envoy/blob/1ae1aa9323fa68bd979da884898444c0471e6fc0/source/common/grpc/async_client_impl.cc#L37

Reply:
Thanks for the confirmation. It makes sense. Seems my previous comments are based on mine incorrect knowledge.